cpp/src/arrow/tensor.cc

```diff
@@ -123,6 +123,80 @@ Type::type Tensor::type_id() const { return type_->id(); }
 
 bool Tensor::Equals(const Tensor& other) const { return TensorEquals(*this, other); }
 
+namespace {
+
+template <typename TYPE>
+size_t StridedTensorCountNonZero(int dim_index, int64_t offset, const Tensor& tensor) {
+  using c_type = typename TYPE::c_type;
+  c_type const zero = c_type(0);
+  size_t nnz = 0;
+  if (dim_index == tensor.ndim() - 1) {
```

Comment:
This needn't be in this PR, but it would be interesting to define a `TensorDataVisitor` (akin to `ArrayDataVisitor`) in `visitor-inline.h`. Bonus points if there's also a `UnorderedTensorDataVisitor` that chooses the optimal iteration order depending on whether the tensor is row-major or column-major ;-)